File: flutter_client/lib/models/department.dart
class Department {
  final String id;
  final String name;
  final String? code;
  final String? description;
  final String? parentId;
  final String? parentName;
  final String? managerId;
  final String? managerName;
  final int? sortOrder;
  final bool isActive;
  final int? employeeCount;
  final int? totalEmployeeCount;
  final int? level;
  final String? hierarchyPath;
  final List<dynamic>? positions;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // Aliases for screen compatibility
  String? get parentDepartmentId => parentId;
  String? get parentDepartmentName => parentName;
  int? get directEmployeeCount => employeeCount;

  Department({
    required this.id,
    required this.name,
    this.code,
    this.description,
    this.parentId,
    this.parentName,
    this.managerId,
    this.managerName,
    this.sortOrder,
    this.isActive = true,
    this.employeeCount,
    this.totalEmployeeCount,
    this.level,
    this.hierarchyPath,
    this.positions,
    this.createdAt,
    this.updatedAt,
  });

  factory Department.fromJson(Map<String, dynamic> json) {
    return Department(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      code: json['code'],
      description: json['description'],
      parentId: json['parentId']?.toString() ?? json['parentDepartmentId']?.toString(),
      parentName: json['parentName'] ?? json['parentDepartmentName'],
      managerId: json['managerId']?.toString(),
      managerName: json['managerName'],
      sortOrder: json['sortOrder'],
      isActive: json['isActive'] ?? true,
      employeeCount: json['employeeCount'] ?? json['directEmployeeCount'],
      totalEmployeeCount: json['totalEmployeeCount'],
      level: json['level'],
      hierarchyPath: json['hierarchyPath'],
      positions: json['positions'] as List<dynamic>?,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'code': code,
    'description': description,
    'parentId': parentId,
    'managerId': managerId,
    'sortOrder': sortOrder,
    'isActive': isActive,
  };
}

class DepartmentTreeNode {
  final String id;
  final String name;
  final String? code;
  final String? parentId;
  final int? level;
  final String? displayName;
  final int? employeeCount;
  final int? totalEmployeeCount;
  final String? managerName;
  final bool isActive;
  final List<DepartmentTreeNode> children;

  int? get directEmployeeCount => employeeCount;

  DepartmentTreeNode({
    required this.id,
    required this.name,
    this.code,
    this.parentId,
    this.level,
    this.displayName,
    this.employeeCount,
    this.totalEmployeeCount,
    this.managerName,
    this.isActive = true,
    this.children = const [],
  });

  factory DepartmentTreeNode.fromJson(Map<String, dynamic> json) {
    return DepartmentTreeNode(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      code: json['code'],
      parentId: json['parentId']?.toString(),
      level: json['level'],
      displayName: json['displayName'],
      employeeCount: json['employeeCount'] ?? json['directEmployeeCount'],
      totalEmployeeCount: json['totalEmployeeCount'],
      managerName: json['managerName'],
      isActive: json['isActive'] ?? true,
      children: (json['children'] as List?)
              ?.map((e) => DepartmentTreeNode.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class DepartmentSelectDto {
  final String id;
  final String name;
  final String? code;
  final int? level;
  final String? displayName;
  final String? parentId;

  DepartmentSelectDto({
    required this.id,
    required this.name,
    this.code,
    this.level,
    this.displayName,
    this.parentId,
  });

  factory DepartmentSelectDto.fromJson(Map<String, dynamic> json) {
    return DepartmentSelectDto(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      code: json['code'],
      level: json['level'],
      displayName: json['displayName'],
      parentId: json['parentId']?.toString(),
    );
  }
}

File: src/ZKTecoADMS.Domain/Entities/Department.cs
using System.ComponentModel.DataAnnotations;
using ZKTecoADMS.Domain.Entities.Base;

namespace ZKTecoADMS.Domain.Entities;

/// <summary>
/// Phòng ban với cấu trúc phân cấp (Hierarchy)
/// </summary>
public class Department : AuditableEntity<Guid>
{
    /// <summary>
    /// Mã phòng ban
    /// </summary>
    [Required]
    [MaxLength(20)]
    public string Code { get; set; } = string.Empty;

    /// <summary>
    /// Tên phòng ban
    /// </summary>
    [Required]
    [MaxLength(200)]
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// Mô tả phòng ban
    /// </summary>
    [MaxLength(500)]
    public string? Description { get; set; }

    /// <summary>
    /// ID phòng ban cha (null nếu là phòng ban gốc)
    /// </summary>
    public Guid? ParentDepartmentId { get; set; }

    /// <summary>
    /// Phòng ban cha
    /// </summary>
    public virtual Department? ParentDepartment { get; set; }

    /// <summary>
    /// Danh sách phòng ban con
    /// </summary>
    public virtual ICollection<Department> Children { get; set; } = new List<Department>();

    /// <summary>
    /// ID người quản lý phòng ban
    /// </summary>
    public Guid? ManagerId { get; set; }

    /// <summary>
    /// Người quản lý phòng ban
    /// </summary>
    public virtual Employee? Manager { get; set; }

    /// <summary>
    /// Cấp độ trong cây phân cấp (0 = gốc, 1 = cấp 1, ...)
    /// </summary>
    public int Level { get; set; }

    /// <summary>
    /// Thứ tự hiển thị trong cùng cấp
    /// </summary>
    public int SortOrder { get; set; }

    /// <summary>
    /// Cửa hàng/Chi nhánh mà phòng ban thuộc về
    /// </summary>
    public Guid? StoreId { get; set; }

    /// <summary>
    /// Cửa hàng/Chi nhánh
    /// </summary>
    public virtual Store? Store { get; set; }

    /// <summary>
    /// Danh sách nhân viên thuộc phòng ban
    /// </summary>
    public virtual ICollection<Employee> Employees { get; set; } = new List<Employee>();

    /// <summary>
    /// Đường dẫn phân cấp (ví dụ: "/root/parent/current/")
    /// Dùng để query nhanh các phòng ban con
    /// </summary>
    [MaxLength(500)]
    public string? HierarchyPath { get; set; }

    /// <summary>
    /// Số lượng nhân viên trực tiếp (cache để tối ưu)
    /// </summary>
    public int DirectEmployeeCount { get; set; }

    /// <summary>
    /// Tổng số nhân viên (bao gồm cả phòng ban con)
    /// </summary>
    public int TotalEmployeeCount { get; set; }

    /// <summary>
    /// Danh sách chức vụ trong phòng ban (JSON array)
    /// </summary>
    [MaxLength(2000)]
    public string? Positions { get; set; }
}
